packages/store/src/external/hydrate.js

```diff
@@ -0,0 +1,113 @@
+// @ts-check
+
+import '../types';
+
+import { makeWeakStore } from '../weak-store';
+import { makeStore } from '../store';
+
+/**
+ * @callback MakeBackingStore
+ * @param {HydrateHook} hydrateHook
+ * @returns {BackingStore}
+ */
+
+/**
+ * This creates an external store maker for a given storage backend, supporting
+ * the Closure interface that the rewriter targets.
+ *
+ * @template {Array<any>} A
+ * @template {ExternalInstance} T
+ * @param {MakeBackingStore} makeBackingStore
+ * @returns {MakeHydrateExternalStore<A, T>}
+ */
+export const makeHydrateExternalStoreMaker = makeBackingStore => {
+  /** @type {WeakStore<T, HydrateKey>} */
+  const instanceToKey = makeWeakStore('instance');
+
+  let lastStoreId = 0;
+
+  // This has to be a strong store, since it is indexed by ID.
+  const storeIdToHydrate = makeStore('storeId');
+
+  /**
+   * Create a data object that queues writes to the store.
+   *
+   * @param {HydrateData} data
+   * @param {() => void} markDirty
+   */
+  const makeActiveData = (data, markDirty) => {
+    const activeData = {};
+    // For every property in data...
+    for (const prop of Object.getOwnPropertyNames(data)) {
+      // Define a getter and setter on activeData.
+      Object.defineProperty(activeData, prop, {
+        get: () => data[prop],
+        set: value => {
+          data[prop] = value;
+          markDirty();
+        },
+      });
+    }
+    return harden(activeData);
+  };
+
+  /** @type {BackingStore} */
+  let backing;
+
+  /** @type {HydrateHook} */
+  const hydrateHook = {
+    getKey(value) {
+      return instanceToKey.get(value);
+    },
+    load([storeId, instanceId]) {
+      const hydrate = storeIdToHydrate.get(storeId);
+      const store = backing.getHydrateStore(storeId);
+
+      const data = store.get(instanceId);
+      const markDirty = () => store.set(instanceId, data);
+
+      const activeData = makeActiveData(data, markDirty);
+      const obj = hydrate(activeData);
+      instanceToKey.init(obj, [storeId, instanceId]);
+      return obj;
+    },
+    drop(storeId) {
+      storeIdToHydrate.delete(storeId);
+    },
+  };
+
+  backing = makeBackingStore(hydrateHook);
+
+  /** @type {MakeHydrateExternalStore<A, T>} */
+  function makeHydrateExternalStore(keyName, adaptArguments, makeHydrate) {
```

Comment:
In all the layering of makers of makers of makers of makers, I'm finding myself completely unable to discern where one injects the actual thing that stores the bits on disk.

Reply:
We should walk through this verbally together and consider alternative names to make the levels clearer. As you know from the E days, ocap style often more heavily curries making functions, because different steps towards instantiation have different authority issues. Better naming can indeed help a lot.